tagless-final/src/main/java/com/github/jorgecastillo/kotlinandroid/data/datasource/remote/MarvelNetworkDataSource.kt

```diff
@@ -12,60 +13,72 @@ import com.karumi.marvelapiclient.MarvelApiException
 import com.karumi.marvelapiclient.MarvelAuthApiException
 import com.karumi.marvelapiclient.model.CharacterDto
 import com.karumi.marvelapiclient.model.CharactersQuery
+import com.karumi.marvelapiclient.model.CharactersQuery.Builder
 import kategory.HK
 import kategory.Option
+import kategory.Try
 import kategory.binding
+import kategory.effects.AsyncContext
+import kategory.right
+import kotlinx.coroutines.experimental.CommonPool
+import kotlinx.coroutines.experimental.async
 import java.net.HttpURLConnection
 
-/*
- * This is the network data source. Calls are made using Karumi's MarvelApiClient.
- * @see "https://github.com/Karumi/MarvelApiClientAndroid"
- *
- * Both requests return a new Reader enclosing an action to resolve when you provide them with the
- * required execution context.
- *
- * The getHeroesFromAvengerComicsUseCase() method maps the fetchAllHeroes() result to filter the list with just the
- * elements with given conditions. It's returning heroes appearing on comics with the  "Avenger"
- * word in the title. Yep, I wanted to retrieve Avengers but the Marvel API is a bit weird
- * sometimes.
- */
-fun exceptionAsCharacterError(e: Throwable): CharacterError =
-    when (e) {
-      is MarvelAuthApiException -> AuthenticationError
-      is MarvelApiException ->
-        if (e.httpCode == HttpURLConnection.HTTP_NOT_FOUND) NotFoundError
-        else UnknownServerError(Option.Some(e))
-      else -> UnknownServerError((Option.Some(e)))
-    }
-
-
 inline fun <reified F> fetchAllHeroes(
     C: MonadControl<F, GetHeroesContext, CharacterError> = monadControl()): HK<F, List<CharacterDto>> =
     C.binding {
-      val query = CharactersQuery.Builder.create().withOffset(0).withLimit(50).build()
+      val query = buildFetchHeroesQuery()
       val ctx = C.ask().bind()
-      C.catch(
-          { ctx.apiClient.getAll(query).response.characters },
-          { exceptionAsCharacterError(it) }
-      )
+      runInAsyncContext(
+          f = { fetchHeroes(ctx, query) },
+          onError = { liftError(C, it) },
+          onSuccess = { liftSuccess(C, it) },
+          AC = C
+      ).bind()
     }
 
 inline fun <reified F> fetchHeroDetails(heroId: String,
-    C: MonadControl<F, GetHeroDetailsContext, CharacterError> = monadControl()): HK<F, CharacterDto> =
+    C: MonadControl<F, GetHeroDetailsContext, CharacterError> = monadControl()): HK<F, List<CharacterDto>> =
```

Comment:
why does this return a List? Can there be more than one hero with the same Id?. I think the correct return type here is `HK<F, CharacterDto>` since we have the notion of `HeroNotFound` which can be raised here `C.raiseError` if we don't get a hero back.